From 8805a8d3f3ccc8a1bcc81f413e0f6ccc31a8c9b2 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Thu, 05 Aug 2021 13:43:07 +0300
Subject: [PATCH] [SW-50880] preboot: fit: move secondary cpu stop function

In secureboot flow, ARC0 is used to run PCIe initialization and
put to sleep once initialization is done. reset needs to be performed
on ARC0 before copying FIT images in ICCM and DCCM memories of ARC.
Currently stop of secondary CPU is performed as part of fit_run_image,
this patch add changes to move it fit_load_images function before
copy of FIT images.

Change-Id: I1e50cd15e3bb83f2799be378c7657b35a4358b36
---

diff --git a/lib/preboot/fit.c b/lib/preboot/fit.c
index 269e159..b9a562f 100644
--- a/lib/preboot/fit.c
+++ b/lib/preboot/fit.c
@@ -939,10 +939,22 @@
 	return serr;
 }
 
+/*
+ * For secureboot flow, ARC0 is used to run
+ * PCIe initialization code and put in sleep
+ * once initialization is done. ARC0 will be
+ * halt and reseted as part of secondary_cpu_stop
+ * function before copying images in ICCM and DCCM
+ * as part of fit_copy_image function.
+ */
+__weak void preboot_secondary_cpu_stop(void) {}
+
 int fit_load_images(void *fit)
 {
 	enum fit_sec_failure sec_res = FIT_SEC_NONE;
 	int err;
+
+	preboot_secondary_cpu_stop();
 
 	memset(&images, 0x0, sizeof(images));
 
@@ -979,7 +991,6 @@
 __weak void preboot_finalize_fw_info(void) {}
 __weak void preboot_finalize_rma_tables(void) {}
 __weak void preboot_finalize_fw_upd_closure(void) {}
-__weak void preboot_secondary_cpu_stop(void) {}
 
 void fit_run_image(void)
 {
@@ -999,7 +1010,6 @@
 	preboot_finalize_rma_tables();
 	preboot_finalize_fw_upd_closure();
 	preboot_finalize_enabled_caps();
-	preboot_secondary_cpu_stop();
 	boot_info_finalize();
 	preboot_start_image(startaddr);
 }
